pymeasure/instruments/lecroy/lecroyT3DSO1204.py

```diff
@@ -0,0 +1,1151 @@
+        will be returned.
+        :param sparsing: interval between data points. For example if sparsing = 4, only one
+        point every 4 points is read.
+        :param averaging: average over data points. For example if averaging = 4, only the 4-point
+        average value is returned.
+        :return: data_ndarray, time_ndarray, waveform_preamble_dict: see waveform_preamble
+        property for dict format. """
+
+        def _header_sanity_checks(message):
+            # The format of the header is DAT2,#9XXXXXXX where XXXXXXX is the number of acquired
+            # points and it is zero padded.
+            message_header = bytes(message[0:self._header_size]).decode("ascii")
+            # Sanity check on header and footer
+            if message_header[0:7] != "DAT2,#9":
+                raise ValueError(f"Waveform data in invalid : header is {message_header}")
+            return int(message_header[-9:])
+
+        def _npoints_sanity_checks(message):
+            message_header = bytes(message[0:self._header_size]).decode("ascii")
+            transmitted_points = int(message_header[-9:])
+            received_points = len(message) - self._header_size - self._footer_size
+            if transmitted_points != received_points:
+                raise ValueError(f"Number of transmitted points ({transmitted_points}) != "
+                                 f"number of received points ({received_points})")
+
+        def _footer_sanity_checks(message):
+            # The footer is always a double line-carriage \n\n
+            message_footer = bytes(message[-self._footer_size:]).decode("ascii")
+            if message_footer != "\n\n":
+                raise ValueError(f"Waveform data in invalid : footer is {message_footer}")
+
+        # First of all set the correct acquisition type
+        if averaging > 1:
+            self.acquisition_type = "average"
+            self.acquisition_average = averaging
+        else:
+            self.acquisition_type = "normal"
+
+        # Check that we are trying to read a reasonable amount of points
+        sample_points = self.acquisition_sample_size(self.waveform_source)
+        if requested_points > 0 and 0 < sample_points < requested_points:
+            raise ValueError(f"Number of requested points ({requested_points}) is greater than "
+                             f"number of sampled points ({sample_points})")
+
+        # Setup waveform acquisition parameters
+        source = _sanitize_source(source)
+        self.waveform_source = source
+        self.waveform_sparsing = sparsing
+        self.waveform_points = requested_points
+        self.waveform_first_point = 0
+
+        # Read waveform preamble (waveform metadata)
+        preamble = self.waveform_preamble
+
+        # Check how many points are to be expected
+        self.write(f"{source}:WF? DAT2")
+        values = self._digitize(source=source)
```

Comment:
Do you on purpose write twice `f"{source}:WF? DAT2"` to the device, as `_digitize` writes it again?

Reply:
No, that is a bug. Thank you for spotting it. Fixed in commit 5f726b7fb41c5014cda61f5f2d3ebd80228477b8